File: microservices/talk-to-observability-agent/app/util/grafana_links.py
from __future__ import annotations

import json
import logging
from datetime import datetime
from functools import lru_cache
from urllib.parse import quote

import httpx

logger = logging.getLogger(__name__)


def _epoch_ms(dt: datetime) -> int:
    return int(dt.timestamp() * 1000)


@lru_cache(maxsize=4)
def resolve_loki_datasource_uid(base_url: str, fallback_uid: str = "loki") -> str:
    return _resolve_datasource_uid(base_url, "Loki", fallback_uid)


@lru_cache(maxsize=4)
def resolve_dashboard_uid(
    base_url: str,
    dashboard_title: str = "Ecommerce Observability",
    fallback_uid: str = "ecommerce-observability",
) -> str:
    url = f"{base_url.rstrip('/')}/api/search"
    try:
        response = httpx.get(
            url,
            params={"query": dashboard_title, "type": "dash-db"},
            timeout=3.0,
        )
        response.raise_for_status()
        for item in response.json():
            if item.get("title") == dashboard_title and item.get("uid"):
                return item["uid"]
    except Exception as exc:
        logger.warning("grafana_dashboard_uid_lookup_failed error=%s", exc)
    return fallback_uid


def _resolve_datasource_uid(base_url: str, name: str, fallback_uid: str) -> str:
    url = f"{base_url.rstrip('/')}/api/datasources/name/{name}"
    try:
        response = httpx.get(url, timeout=3.0)
        response.raise_for_status()
        uid = response.json().get("uid")
        if uid:
            return uid
    except Exception as exc:
        logger.warning("grafana_datasource_uid_lookup_failed name=%s error=%s", name, exc)
    return fallback_uid


def build_loki_explore_url(
    base_url: str,
    start: datetime,
    end: datetime,
    *,
    correlation_id: str | None = None,
    fallback_loki_uid: str = "loki",
    api_base_url: str | None = None,
) -> str:
    uid = resolve_loki_datasource_uid(api_base_url or base_url, fallback_loki_uid)
    if correlation_id:
        expr = f'{{namespace=~"ecommerce|observability"}} |= "{correlation_id}"'
    else:
        expr = '{namespace=~"ecommerce|observability"}'

    pane = {
        "datasource": uid,
        "queries": [
            {
                "refId": "A",
                "expr": expr,
                "queryType": "range",
                "datasource": {"type": "loki", "uid": uid},
            }
        ],
        "range": {"from": str(_epoch_ms(start)), "to": str(_epoch_ms(end))},
    }
    panes = {"a": pane}
    encoded = quote(json.dumps(panes, separators=(",", ":")), safe="")
    return f"{base_url.rstrip('/')}/explore?orgId=1&panes={encoded}&schemaVersion=1"


def build_dashboard_url(
    base_url: str,
    start: datetime,
    end: datetime,
    *,
    fallback_dashboard_uid: str = "ecommerce-observability",
    api_base_url: str | None = None,
) -> str:
    uid = resolve_dashboard_uid(
        api_base_url or base_url, fallback_uid=fallback_dashboard_uid
    )
    return (
        f"{base_url.rstrip('/')}/d/{uid}"
        f"?from={_epoch_ms(start)}&to={_epoch_ms(end)}&timezone=UTC"
    )

File: microservices/observability-debug-agent-java/src/main/java/com/observability/agent/util/GrafanaLinks.java
package com.observability.agent.util;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.observability.agent.config.GrafanaProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.beans.factory.annotation.Qualifier;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Builds Grafana Explore (Loki) and Dashboard URLs.
 * Mirrors Python app/util/grafana_links.py.
 */
@Component
public class GrafanaLinks {

    private static final Logger log = LoggerFactory.getLogger(GrafanaLinks.class);

    private final GrafanaProperties grafanaProps;
    private final WebClient grafanaWebClient;
    private final ObjectMapper mapper = new ObjectMapper();
    private final Map<String, String> uidCache = new ConcurrentHashMap<>();

    public GrafanaLinks(GrafanaProperties grafanaProps,
                        @Qualifier("grafanaWebClient") WebClient grafanaWebClient) {
        this.grafanaProps = grafanaProps;
        this.grafanaWebClient = grafanaWebClient;
    }

    public String buildLokiExploreUrl(String correlationId, String namespace,
                                      String startTime, String endTime) {
        String datasourceUid = resolveLokiDatasourceUid();
        String logql = correlationId != null && !correlationId.isEmpty()
                ? String.format("{correlationId=\"%s\"}", correlationId)
                : String.format("{namespace=\"%s\"}", namespace);

        try {
            long fromMs = toEpochMs(startTime);
            long toMs   = toEpochMs(endTime);

            Map<String, Object> pane = Map.of(
                    "queries", new Object[]{Map.of(
                            "refId", "A",
                            "expr", logql,
                            "datasource", Map.of("type", "loki", "uid", datasourceUid)
                    )},
                    "range", Map.of("from", String.valueOf(fromMs), "to", String.valueOf(toMs))
            );

            String panesJson = mapper.writeValueAsString(Map.of("a", pane));
            String encoded = URLEncoder.encode(panesJson, StandardCharsets.UTF_8);
            return grafanaProps.getBaseUrl() + "/explore?panes=" + encoded + "&schemaVersion=1&orgId=1";
        } catch (Exception e) {
            log.warn("Failed to build Loki explore URL: {}", e.getMessage());
            return null;
        }
    }

    public String buildDashboardUrl(String startTime, String endTime) {
        String dashUid = resolveDashboardUid();
        try {
            long fromMs = toEpochMs(startTime);
            long toMs   = toEpochMs(endTime);
            return grafanaProps.getBaseUrl()
                    + "/d/" + dashUid
                    + "?orgId=1&from=" + fromMs + "&to=" + toMs;
        } catch (Exception e) {
            log.warn("Failed to build dashboard URL: {}", e.getMessage());
            return null;
        }
    }

    private String resolveLokiDatasourceUid() {
        return uidCache.computeIfAbsent("loki", k -> {
            try {
                var node = grafanaWebClient.get()
                        .uri("/api/datasources/name/Loki")
                        .retrieve()
                        .bodyToMono(com.fasterxml.jackson.databind.JsonNode.class)
                        .block();
                return node != null ? node.path("uid").asText(grafanaProps.getLokiDatasourceUid())
                                    : grafanaProps.getLokiDatasourceUid();
            } catch (Exception e) {
                return grafanaProps.getLokiDatasourceUid();
            }
        });
    }

    private String resolveDashboardUid() {
        return uidCache.computeIfAbsent("dashboard", k -> {
            try {
                var nodes = grafanaWebClient.get()
                        .uri("/api/search?type=dash-db")
                        .retrieve()
                        .bodyToMono(com.fasterxml.jackson.databind.JsonNode.class)
                        .block();
                if (nodes != null && nodes.isArray()) {
                    for (var n : nodes) {
                        if ("ecommerce-observability".equalsIgnoreCase(n.path("title").asText())) {
                            return n.path("uid").asText(grafanaProps.getDashboardUid());
                        }
                    }
                }
                return grafanaProps.getDashboardUid();
            } catch (Exception e) {
                return grafanaProps.getDashboardUid();
            }
        });
    }

    private long toEpochMs(String isoTime) {
        try {
            return Instant.parse(isoTime).toEpochMilli();
        } catch (DateTimeParseException e) {
            return Instant.now().toEpochMilli();
        }
    }
}
